### DISCUSSION TITLE
Modelling question: active users

### DISCUSSION THREAD

**Discussion by srijken** [ravendb/ravendb#13381] [Q&A] [answered]
We want to keep a record of when users were active, an be able to view this data yearly, quarterly etc. What would be the best way to model this? Is time series suitable for this, or do I just have a 'user' document with a list of activity/login dates?

Thanks,
Sander

**Comment by ayende** [comment, marked as answer]
Sounds like a really great use for time series, no?
See: https://ravendb.net/articles/ravendb-5-3-features-incremental-time-series-throttling-and-rate-limits

**Reply by srijken** [reply]
This is starting to come along, but I have some additional questions.

```
    public class User
    {
        public string Id { get; set; }
        public string Tenant { get; set; }
    }
```

every time a user logs on I do this:
```
                session.IncrementalTimeSeriesFor(user, "INC:UserLogon")
                    .Increment(1);

                await session.SaveChangesAsync(cancellationToken).ConfigureAwait(false);
```

the question now: how can I most efficiently get a unique user per time period result.

At the moment I have:
```
                var result = await session.Query<User>()
                    .Where(x => x.Tenant == tenant, false)
                    .Select(x => RavenQuery.TimeSeries(x, "INC:UserLogon", from, to)
                        .GroupBy(x => x.Days(1)) // maybe change this to something more sensible
                        .Select(x => x.Min())
                        .ToList()
                    )
                    .ToListAsync(cancellationToken).ConfigureAwait(false);
```

the problem is that `result` now contains an entry for each 'user'. I'd like a total of users that had '1' logon in the selected time period.

**Reply by ayende** [reply]
The easiest way to do that is to have multiple time series, one for each user and one for the tenant. That would give you the details that you want.
Otherwise, you'll need to do a map/reduce index on the time series and aggregate on per tenant, but separate time series is probably best.

**Reply by srijken** [reply]
One time series per tenant can't work (or I misunderstood you), because you don't know if a user has been counted yet or not without storing that per user. And because the time window for the query can vary, you need to store each logon attempt

**Reply by ayende** [reply]
You can use:

                session.IncrementalTimeSeriesFor(user, "INC:UserLogon")
                    .Increment(1);

                session.IncrementalTimeSeriesFor(user.Tenant, "INC:UserLogon")
                    .Increment(1);

                await session.SaveChangesAsync(cancellationToken).ConfigureAwait(false);


And then you can issue the query at the tenant level and get the results. You are recording this at both the user and tenant levels.

**Reply by srijken** [reply]
suppose just one user logs on 3 times during a day. Then both incremental time series for that day are at '3'. I want to know that it was just one user. I don't see this working with the user.Tenant construction?

**Reply by ayende** [reply]
Oh, you want total number of _unique_ users on a daily basis?
Then you should probably use the map/reduce in that case, yes.

**Reply by srijken** [reply]
OK, I'll try that direction!

**Reply by srijken** [reply]
Are there any time-series - map/reduce examples? I was trying this, and I didn't figure out a way where I can modify the from/to afterwards, because the index has already grouped per time window. Is there a way to fix this, or should I do an index per month/year/etc?

**Reply by ayende** [reply]
The group by is usually by a fixed range (day, month, etc).
And you can aggregate on top of that if you need to get the numbers on a bigger range.
So you can do a group by day, then aggregate to the month using facets. 

https://ravendb.net/docs/article-page/5.2/NodeJs/document-extensions/timeseries/indexing

**Reply by srijken** [reply]
The problem with that is that it's not possible to simply add the daily counts together. One user might be active every day, whereas another user might be active only one day in the month.

It would be helpful if it was possible to filter the time series results on Count > 0, and aggregate the data in some way on the server.

**Reply by ayende** [reply]
@srijken I'm not sure that I understand what you mean?

You are recording the logins each time a user is logging in.
And then what queries do you want to run?

Unique users by tenant on a daily basis? Unique users by date range?

Let's assume that you have the following index:

```
// map
from ts in timeseries.Users.Logins
let user = LoadDocument("Users", ts.DocumentId)
from entry in ts.Entries
select new 
{
    entry.Timestamp,
    User = ts.DocumentId,
    user.TenantId
}
```

That index all the relevant logins by user and tenant.

Then you can query this using:

```
from index 'Logins/Times'
where TenantId = $tenantId and Timestamp between $start and $end
select facet(User)
```


That should give you the list of unique users that were logged in during that time frame.

**Reply by srijken** [reply]
@ayende That does the trick! Sorry for taking a while to grasp the new concept :)
